packages/ckeditor5-widget/tests/widgettypearound/widgettypearound.js

```diff
@@ -0,0 +1,345 @@
+
+		it( 'should detect widgets that are a last child of the parent container', () => {
+			setModelData( editor.model, '<paragraph>foo</paragraph><blockWidget></blockWidget>' );
+
+			const viewWidget = viewRoot.getChild( 1 );
+
+			assertIsTypeAroundBefore( viewWidget, false );
+			assertIsTypeAroundAfter( viewWidget, true );
+		} );
+
+		it( 'should not detect widgets that are surrounded by sibling which allow the selection', () => {
+			setModelData( editor.model, '<paragraph>foo</paragraph><blockWidget></blockWidget><paragraph>bar</paragraph>' );
+
+			const viewWidget = viewRoot.getChild( 1 );
+
+			assertIsTypeAroundBefore( viewWidget, false );
+			assertIsTypeAroundAfter( viewWidget, false );
+		} );
+
+		it( 'should detect widgets that have another block widget as a next sibling', () => {
+			setModelData( editor.model, '<blockWidget></blockWidget><blockWidget></blockWidget>' );
+
+			const firstViewWidget = viewRoot.getChild( 0 );
+
+			assertIsTypeAroundBefore( firstViewWidget, true );
+			assertIsTypeAroundAfter( firstViewWidget, true );
+		} );
+
+		it( 'should detect widgets that have another block widget as a previous sibling', () => {
+			setModelData( editor.model, '<blockWidget></blockWidget><blockWidget></blockWidget>' );
+
+			const lastViewWidget = viewRoot.getChild( 1 );
+
+			assertIsTypeAroundBefore( lastViewWidget, true );
+			assertIsTypeAroundAfter( lastViewWidget, true );
+		} );
+
+		it( 'should not detect inline widgets even if they fall in previous categories', () => {
+			setModelData( editor.model,
+				'<paragraph><inlineWidget></inlineWidget><inlineWidget></inlineWidget></paragraph>'
+			);
+
+			const firstViewWidget = viewRoot.getChild( 0 ).getChild( 0 );
+			const lastViewWidget = viewRoot.getChild( 0 ).getChild( 1 );
+
+			assertIsTypeAroundBefore( firstViewWidget, false );
+			assertIsTypeAroundAfter( firstViewWidget, false );
+
+			assertIsTypeAroundBefore( lastViewWidget, false );
+			assertIsTypeAroundAfter( lastViewWidget, false );
+		} );
+	} );
+
+	function assertIsTypeAroundBefore( viewWidget, expected ) {
+		expect( viewWidget.hasClass( 'ck-widget_can-type-around_before' ) ).to.equal( expected );
+	}
+
+	function assertIsTypeAroundAfter( viewWidget, expected ) {
+		expect( viewWidget.hasClass( 'ck-widget_can-type-around_after' ) ).to.equal( expected );
+	}
+
+	function blockWidgetPlugin( editor ) {
+		editor.model.schema.register( 'blockWidget', {
+			inheritAllFrom: '$block',
+			isObject: true
+		} );
+
+		editor.conversion.for( 'downcast' )
+			.elementToElement( {
+				model: 'blockWidget',
+				view: ( modelItem, viewWriter ) => {
+					const container = viewWriter.createContainerElement( 'div' );
+					const viewText = viewWriter.createText( 'block-widget' );
+
+					viewWriter.insert( viewWriter.createPositionAt( container, 0 ), viewText );
+
+					return toWidget( container, viewWriter, {
+						label: 'block widget'
+					} );
+				}
+			} );
+	}
+
+	function inlineWidgetPlugin( editor ) {
```

Comment:
Maybe this could be extracted to some util?

Reply:
See #6813.